File: frontend/src/app/core/services/post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Post {
  id: number;
  user: any;
  title: string;
  description?: string;
  authorName?: string;
  bookTitle?: string;
  genre?: string;
  price: number;
  condition: string;
  city: string;
  area?: string;
  contactPhone?: string;
  contactEmail?: string;
  images: string[];
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface PostRequest {
  title: string;
  description?: string;
  authorName?: string;
  bookTitle?: string;
  genre?: string;
  price: number;
  condition: string;
  city: string;
  area?: string;
  contactPhone?: string;
  contactEmail?: string;
  images: string[];
}

export interface PageResponse<T> {
  content: T[];
  pageable: any;
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
}

@Injectable({
  providedIn: 'root'
})
export class PostService {
  private readonly API_URL = `${environment.apiUrl}/posts`;

  constructor(private http: HttpClient) {}

  getAllPosts(page: number = 0, size: number = 20, sortBy: string = 'createdAt', sortDir: string = 'DESC'): Observable<PageResponse<Post>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    return this.http.get<PageResponse<Post>>(this.API_URL, { params });
  }

  getPosts(page: number = 0, size: number = 20, sortBy: string = 'createdAt', sortDir: string = 'DESC'): Observable<PageResponse<Post>> {
    return this.getAllPosts(page, size, sortBy, sortDir);
  }

  searchPosts(
    query?: string, 
    genre?: string, 
    city?: string, 
    minPrice?: number | null, 
    maxPrice?: number | null,
    condition?: string,
    page: number = 0, 
    size: number = 20,
    sortBy: string = 'createdAt',
    sortDir: string = 'DESC'
  ): Observable<PageResponse<Post>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    if (query) params = params.set('query', query);
    if (genre) params = params.set('genre', genre);
    if (city) params = params.set('city', city);
    if (minPrice !== null && minPrice !== undefined) params = params.set('minPrice', minPrice.toString());
    if (maxPrice !== null && maxPrice !== undefined) params = params.set('maxPrice', maxPrice.toString());
    if (condition) params = params.set('condition', condition);

    return this.http.get<PageResponse<Post>>(`${this.API_URL}/search`, { params });
  }

  getFeaturedPosts(): Observable<Post[]> {
    return this.http.get<Post[]>(`${this.API_URL}/featured`);
  }

  getPostById(id: number): Observable<Post> {
    return this.http.get<Post>(`${this.API_URL}/${id}`);
  }

  getPost(id: number): Observable<Post> {
    return this.getPostById(id);
  }

  createPost(request: PostRequest): Observable<Post> {
    return this.http.post<Post>(this.API_URL, request);
  }

  updatePost(id: number, request: PostRequest): Observable<Post> {
    return this.http.put<Post>(`${this.API_URL}/${id}`, request);
  }

  deletePost(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`);
  }

  getUserPosts(userId: number, page: number = 0, size: number = 20): Observable<PageResponse<Post>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<PageResponse<Post>>(`${this.API_URL}/user/${userId}`, { params });
  }
}

File: backend/src/main/java/com/kitabqurdu/service/PostService.java
package com.kitabqurdu.service;

import com.kitabqurdu.dto.PostDto;
import com.kitabqurdu.dto.PostRequest;
import com.kitabqurdu.dto.UserDto;
import com.kitabqurdu.model.Post;
import com.kitabqurdu.model.User;
import com.kitabqurdu.repository.PostRepository;
import com.kitabqurdu.repository.UserRepository;
import com.kitabqurdu.security.UserPrincipal;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class PostService {

    @Autowired
    private PostRepository postRepository;

    @Autowired
    private UserRepository userRepository;

    @Cacheable(value = "posts", key = "#pageable.pageNumber + '-' + #pageable.pageSize")
    public Page<PostDto> getAllPosts(Pageable pageable) {
        return postRepository.findByStatus(Post.Status.ACTIVE, pageable)
                .map(this::convertToDto);
    }

    public Page<PostDto> searchPosts(String query, String genre, String city, 
                                     BigDecimal minPrice, BigDecimal maxPrice, 
                                     String condition, Pageable pageable) {
        Specification<Post> spec = Specification.where(null);

        spec = spec.and((root, cq, cb) -> cb.equal(root.get("status"), Post.Status.ACTIVE));

        if (query != null && !query.isEmpty()) {
            spec = spec.and((root, cq, cb) -> cb.or(
                cb.like(cb.lower(root.get("title")), "%" + query.toLowerCase() + "%"),
                cb.like(cb.lower(root.get("description")), "%" + query.toLowerCase() + "%"),
                cb.like(cb.lower(root.get("authorName")), "%" + query.toLowerCase() + "%"),
                cb.like(cb.lower(root.get("bookTitle")), "%" + query.toLowerCase() + "%")
            ));
        }

        if (genre != null && !genre.isEmpty()) {
            spec = spec.and((root, cq, cb) -> cb.equal(root.get("genre"), genre));
        }

        if (city != null && !city.isEmpty()) {
            spec = spec.and((root, cq, cb) -> cb.equal(root.get("city"), city));
        }

        if (minPrice != null) {
            spec = spec.and((root, cq, cb) -> cb.greaterThanOrEqualTo(root.get("price"), minPrice));
        }

        if (maxPrice != null) {
            spec = spec.and((root, cq, cb) -> cb.lessThanOrEqualTo(root.get("price"), maxPrice));
        }

        if (condition != null && !condition.isEmpty()) {
            spec = spec.and((root, cq, cb) -> cb.equal(root.get("condition"), Post.Condition.valueOf(condition)));
        }

        return postRepository.findAll(spec, pageable).map(this::convertToDto);
    }

    public PostDto getPostById(Long id) {
        Post post = postRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Post not found"));
        return convertToDto(post);
    }

    @Transactional
    public PostDto createPost(PostRequest request, Authentication authentication) {
        UserPrincipal userPrincipal = (UserPrincipal) authentication.getPrincipal();
        User user = userRepository.findById(userPrincipal.getId())
                .orElseThrow(() -> new RuntimeException("User not found"));

        Post post = new Post();
        post.setUser(user);
        post.setTitle(request.getTitle());
        post.setDescription(request.getDescription());
        post.setAuthorName(request.getAuthorName());
        post.setBookTitle(request.getBookTitle());
        post.setGenre(request.getGenre());
        post.setPrice(request.getPrice());
        post.setCondition(Post.Condition.valueOf(request.getCondition()));
        post.setCity(request.getCity());
        post.setArea(request.getArea());
        post.setContactPhone(request.getContactPhone());
        post.setContactEmail(request.getContactEmail());
        post.setImages(request.getImages());
        post.setStatus(Post.Status.ACTIVE);

        post = postRepository.save(post);
        return convertToDto(post);
    }

    @Transactional
    public PostDto updatePost(Long id, PostRequest request, Authentication authentication) {
        UserPrincipal userPrincipal = (UserPrincipal) authentication.getPrincipal();
        Post post = postRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getUser().getId().equals(userPrincipal.getId())) {
            throw new RuntimeException("You don't have permission to update this post");
        }

        post.setTitle(request.getTitle());
        post.setDescription(request.getDescription());
        post.setAuthorName(request.getAuthorName());
        post.setBookTitle(request.getBookTitle());
        post.setGenre(request.getGenre());
        post.setPrice(request.getPrice());
        post.setCondition(Post.Condition.valueOf(request.getCondition()));
        post.setCity(request.getCity());
        post.setArea(request.getArea());
        post.setContactPhone(request.getContactPhone());
        post.setContactEmail(request.getContactEmail());
        post.setImages(request.getImages());

        post = postRepository.save(post);
        return convertToDto(post);
    }

    @Transactional
    public void deletePost(Long id, Authentication authentication) {
        UserPrincipal userPrincipal = (UserPrincipal) authentication.getPrincipal();
        Post post = postRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Post not found"));

        if (!post.getUser().getId().equals(userPrincipal.getId())) {
            throw new RuntimeException("You don't have permission to delete this post");
        }

        post.setStatus(Post.Status.ARCHIVED);
        postRepository.save(post);
    }

    public Page<PostDto> getUserPosts(Long userId, Pageable pageable) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return postRepository.findByUser(user, pageable).map(this::convertToDto);
    }

    public List<PostDto> getFeaturedPosts() {
        return postRepository.findTop10ByStatusOrderByCreatedAtDesc(Post.Status.ACTIVE)
                .stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    private PostDto convertToDto(Post post) {
        PostDto dto = new PostDto();
        dto.setId(post.getId());
        dto.setUser(convertUserToDto(post.getUser()));
        dto.setTitle(post.getTitle());
        dto.setDescription(post.getDescription());
        dto.setAuthorName(post.getAuthorName());
        dto.setBookTitle(post.getBookTitle());
        dto.setGenre(post.getGenre());
        dto.setPrice(post.getPrice());
        dto.setCondition(post.getCondition().name());
        dto.setCity(post.getCity());
        dto.setArea(post.getArea());
        dto.setContactPhone(post.getContactPhone());
        dto.setContactEmail(post.getContactEmail());
        dto.setImages(post.getImages());
        dto.setStatus(post.getStatus().name());
        dto.setCreatedAt(post.getCreatedAt());
        dto.setUpdatedAt(post.getUpdatedAt());
        return dto;
    }

    private UserDto convertUserToDto(User user) {
        UserDto dto = new UserDto();
        dto.setId(user.getId());
        dto.setUsername(user.getUsername());
        dto.setEmail(user.getEmail());
        dto.setFirstName(user.getFirstName());
        dto.setLastName(user.getLastName());
        dto.setImageUrl(user.getImageUrl());
        return dto;
    }
}
